financial-templates-lib/ExpiringMultiPartyEventClient.js

```diff
@@ -0,0 +1,168 @@
+// a clear separation of concerns and to limit the overhead from querying chain necessarily.
+// If no updateThreshold is specified then default to updating every 60 seconds.
+class ExpiringMultiPartyEventClient {
+  constructor(abi, web3, empAddress, updateThreshold = 60) {
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+
+    this.web3 = web3;
+
+    // EMP contract
+    this.emp = new web3.eth.Contract(abi, empAddress);
+    this.empAddress = empAddress;
+
+    // EMP Events
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  }
+
+  // Calls _update unless it was recently called, as determined by this.updateThreshold.
+  update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime < this.lastUpdateTimestamp + this.updateThreshold) {
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state update skipped",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTimestamp,
+        timeRemainingUntilUpdate: this.lastUpdateTimestamp + this.updateThreshold - currentTime
+      });
+      return;
+    } else {
+      await this._update();
+      this.lastUpdateTimestamp = currentTime;
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state updated",
+        lastUpdateTimestamp: this.lastUpdateTimestamp
+      });
+    }
+  };
+
+  // Force call of _update, designed to be called by downstream caller that knowingly updated the EMP state.
+  forceUpdate = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    await this._update();
+    this.lastUpdateTimestamp = currentTime;
+    Logger.debug({
+      at: "ExpiringMultiPartyEventClient",
+      message: "EMP state force updated",
+      lastUpdateTimestamp: this.lastUpdateTimestamp
+    });
+  };
+
+  // Delete all events within the client
+  clearState = async () => {
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  };
+
+  // Returns an array of liquidation events.
+  getAllLiquidationEvents = () => this.liquidationEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeEvents = () => this.disputeEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeSettlementEvents = () => this.disputeSettlementEvents;
+
+  start = () => {
+    this._poll();
+  };
+
+  _poll = async () => {
+    while (true) {
+      try {
+        await this._update();
+      } catch (error) {
+        Logger.error({
+          at: "ExpiringMultiPartyEventClient",
+          message: "client polling error",
+          error: error
+        });
+      }
+      await delay(Number(10_000));
+    }
+  };
+
+  _update = async () => {
+    //Events
+    console.log("this.liquidationEvents", this.liquidationEvents);
+    let liquidationBlockQueryNumber =
```

Comment:
An alternative way to do this: query the latest block number, make all 3 of your block ranges from the last run's latest block number to this round's latest block number. This way, you wouldn't have to figure out the block number in three different places, and the queries would be over a smaller range.

To decrease the chance of reorgs causing issues (both in your current implementation and ^that alternative), you *could* consider only querying up to `n` blocks before the latest block, so the state you're persisting is (almost) guaranteed to be stable. The downside is that this makes the notifications lag.

Reply:
This is a good idea, thanks! will implement this and push the appropriate changes.